Added in clicks to keymap.

adamharrison opened the pull request:
Simple change that just adds the ability to bind click events to commands. This seems like a logical, and "least-surprise" extension to the keymap.

And of course, this is an *extension* of the current system, so all existing plugins should work fine. And it comes with the bonus that there's no actual real extra lines of code; things are actually decently cleaner this way.

adamharrison commented:
Should close #558

adamharrison commented:
I think we should probably also introduce the concept of a double and triple click, which would probably clean this up more. Will do that.

Guldoman commented on data/core/view.lua:
```diff
@@ -81,6 +81,7 @@ function View:on_mouse_pressed(button, x, y, clicks)
     self.dragging_scrollbar = true
     return true
   end
+  return keymap.on_key_pressed(button:sub(1,1) .. "click", x, y, clicks)
```
I think it would be better to do something like this in `core.on_event`:
```lua
elseif type == "mousepressed" then
    did_keymap = keymap.on_key_pressed(...) -- TODO: Translate the event name appropriately
    if not did_keymap then
        core.root_view:on_mouse_pressed(...)
    end
```
Also we should make `src/api/system.c:button_name` return values for other mouse buttons, as now we just return `?`.

Do we need to do this also for `mousereleased`?

adamharrison replied:
> I think it would be better to do something like this in core.on_event:

Yeah, agreed, I'll change it. I originally did it in view, because I wanted to make the views easily able to override this new behaviour, but we should keep the same paradigm across keymaps in the long run.

> Also we should make src/api/system.c:button_name return values for other mouse buttons, as now we just return ?.

Sure, I'll add in X1 and X2?

> Do we need to do this also for mousereleased?

No. Keymap doesn't do this, so let's keep it as keymap is.

kivutar commented:
I compiled your code and tried adding this to my user module:

```
keymap.add {
  ["cmd+lclick"] = "lsp:goto-definition",
}
```

It's somehow working, but does not relocate the cursor soon enough to jump to the right definition. It's jumping where the cursor used to be before I perform the keymap.

adamharrison commented:
> It's somehow working, but does not relocate the cursor soon enough to jump to the right definition. It's jumping where the cursor used to be before I perform the keymap.

I noticed a lot of modules don't follow the convention of actually reporting when they actually do something with user input, so this could be related to something like that. We probably just need to `return true` somewhere where we're not currently doing that. I'll look into it.

Guldoman commented:
> I noticed a lot of modules don't follow the convention of actually reporting when they actually do something with user input, so this could be related to something like that. We probably just need to `return true` somewhere where we're not currently doing that. I'll look into it.

Could this be solved by doing first `keymap.on_mouse_pressed` and then `core.root_view:on_mouse_pressed`?

kivutar commented:
Even with this current limitation that seem to affect only some plugins, this PR still makes the code better and more powerful.

adamharrison commented:
OK, so added in mouse wheel into this as well (to simplify the scale plugin), and generalized clicks so that there's `s`, `d` and `t` prefixes for single, double and triple clicking; and a generalized `click` that just handles any amount of clicks which gets called last.

I think this is good to go. If no one has objections, I'd like to merge this by the 12th.

takase1121 commented:
I wonder if instead of `s` `d` `t` we can use a number, like `123dclick`

adamharrison commented:
... That makes way more sense. OK. Will change that.

adamharrison commented:
OK; I've incorporated @takase1121 's suggestions; now we use numeric prefixes (which makes way more sense); as well as allow for both optional click numbers, and optional button specifiers (and renamed x2 to y).

I'll merge this by the 16th if no further comments.
